### ISSUE TITLE
Change Request: More flexible languageOptions serialization

### ISSUE THREAD

**Issue by nzakas** [eslint/eslint#19759]
### ESLint version

HEAD

### What problem do you want to solve?

Currently, when we serialize `languageOptions`, and non-serializable value throws an error. For instance:

```js
{
    languageOptions: {
        parse() {}
    }
}
```

We work around this for parsers by not attempting to serialize an object if we can calculate an object ID for it (due to presence of `meta` property). However, it's not practical add `meta` on any object that might be passed into `languageOptions`, especially if that object is deeply nested. This leaves users and language plugin authors a bit stuck if they need a function to be passed in.

### What do you think is the correct solution?

We can modify the current serialization logic to also check if a value has a `toJSON()` method (which is the standard way to indicate a different JSON serialization behavior for any object), and if so, call `toJSON()` instead of trying to serialize the object.

This would allow language plugin authors to place `toJSON()` on any object that can't be natively serialized and would allow more functionality to be available through `languageOptions`.

### Participation

- [x] I am willing to submit a pull request for this change.

**Comment by mdjermanovic** [comment]
> This would allow language plugin authors to place `toJSON()` on any object that can't be natively serialized

Language plugins would do this in the `normalizeLanguageOptions()` method?

**Comment by mdjermanovic** [comment]
> This leaves users and language plugin authors a bit stuck if they need a function to be passed in.

How would functions be serialized by a custom `.toJSON()`?

**Comment by nzakas** [comment]
> > This would allow language plugin authors to place `toJSON()` on any object that can't be natively serialized
> 
> Language plugins would do this in the `normalizeLanguageOptions()` method?

I hadn't thought of that and that's a good idea! :+1:

> > This leaves users and language plugin authors a bit stuck if they need a function to be passed in.
> 
> How would functions be serialized by a custom `.toJSON()`?

Not the functions, the objects they belong to. So basically, when a value is an object we currently check to see if it has `meta` info, and if so, we don't serialize the object and instead use the name and version (like `parser@1.0.1`). What I'm proposing here is if `meta` is not present, we then check to see if `toJSON()` is present, and if so, call that. Here's the draft PR:
https://github.com/eslint/eslint/pull/19760/files#diff-260c79e7a2f3cfd45a6b348b67610352f644ca1c2abf0ce34c1edaa47019848bR281-R290

**Comment by mdjermanovic** [comment]
> > > This leaves users and language plugin authors a bit stuck if they need a function to be passed in.
> > 
> > 
> > How would functions be serialized by a custom `.toJSON()`?
> 
> Not the functions, the objects they belong to. So basically, when a value is an object we currently check to see if it has `meta` info, and if so, we don't serialize the object and instead use the name and version (like `parser@1.0.1`). What I'm proposing here is if `meta` is not present, we then check to see if `toJSON()` is present, and if so, call that. Here's the draft PR: https://github.com/eslint/eslint/pull/19760/files#diff-260c79e7a2f3cfd45a6b348b67610352f644ca1c2abf0ce34c1edaa47019848bR281-R290

I got it, but how would the custom `toJSON()` handle functions? Would they be just omitted from the output?

**Comment by nzakas** [comment]
> I got it, but how would the custom `toJSON()` handle functions? Would they be just omitted from the output?

That would be up to the language itself to decide. It could just to simply omit functions, or to replace the entire object with a string identifier (like we do when `meta` is present), or convert the function to a boolean, etc. Bottom line: it would be up to the language to attach `toJSON()` with the desired behavior.

**Comment by mdjermanovic** [comment]
Makes sense to me 👍
